### PR TITLE
Add settings for Stork input and output options

### LINKED ISSUE #4: add other options to stork.toml
- [x] I have searched the [issues](https://github.com/pelican-plugins/search/issues?q=is%3Aissue) (including closed ones) and believe that this is not a duplicate.
- [x] I have searched the documentation and believe that my question is not covered.
- [x] I am willing to lend a hand to help implement this feature. 

## Feature Request
Stork has some other options in stork.toml to control index like `stemming` or `url_prefix`. It is needed if your pelican content is not written in English or the site has a certain structure.

## Implementation options

1. pipe all current settings to stork in a one to one matter like `STORK_URL_PREFIX` will be `url_prefix` in input
2. add one setting like `STORK_INPUT_OPTIONS` to have all settings in one place.

The benefit of the latter is that we do not need to change anything, if stork add new options. But it is perhaps a bit more complex from a user perspective and easier to mess up.

I'm willing to implement it, but just wanted to discuss options first

### PR DESCRIPTION
**Pull request by lioman**
Stork index creation can be configured via [additional options](https://stork-search.net/docs/config-ref).
To allow all of them add SEARCH_ADDITIONAL_OPTIONS settings.
Content will be added to stork.toml in input block.

# Pull Request Checklist

Resolves: #4  

- [x] Conformed to **code style guidelines** by running appropriate linting tools
- [x] Updated **documentation** for changed code

### CODE REVIEW

**Comment by lioman** [comment]
Do we need anything to get this merged? I don't want to pin this dependency to git but use a normal version instead

**Comment by justinmayer** [comment]
Hey Lioman. Anything that adds/changes settings has to be considered a bit more carefully than other contributions, since changing them later can cause problems for people using previous setting names. I am completely swamped at the moment, so all I can promise is that I will review this in depth as soon as I possibly can.

**Comment by lioman** [comment]
Sure! No problem, thanks for answering. I added as additional parameter. So if you do not add it, nothing changes.

**Comment by lioman** [comment]
Sorry to ping you again. @justinmayer Is it possible, that you review this? I hate long pending tasks

**Comment by Kristinita** [comment]
**Type: Request** :mailbox_with_mail:

### 1. Request

@lioman , could you make that Pelican users add options in the traditional Pelican format, which is used in other plugins?

Example:

```python
SEARCH_INPUT_OPTIONS = {
	"displayed_results_count": 100
}
```

### 2. Argumentation

[**I store my Pelican settings in `.yaml`**](https://github.com/Kristinita/KristinitaPelican/blob/master/pelican-config/pelican_plugins_settings.yaml), not `.py` files. YAML is a specifically configuration file format, and [**I don’t know**](https://stackoverflow.com/q/49901577/5951529) how to parse the configuration in Python files directly.

If the settings after your change could be entered in the traditional format, YAML users could add them like this:

```yaml
SEARCH_INPUT_OPTIONS:
  displayed_results_count: 100
```

I don’t know, how to convert [**your chosen option format**](https://github.com/pelican-plugins/search/pull/5/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R68) to YAML.

Thanks.

**Comment by lioman** [comment]
@Kristinita I can, but this would be way more complicated, as the settings need to be parsed and translated to a toml format. I chose the current because it is the simplest approach. Basically, it is just a string the gets directly injected and for your yaml setting this should be possible too. Have not tested it, but this should be valid
```yaml
SEARCH_INPUT_OPTIONS: |
    stemming = "German"
    url_prefix = "https://example.com"
```

**Comment by Kristinita** [comment]
**Type: Another request** :mailbox_with_mail:

Thanks, it works for me!

But when I try to change the value of the option `displayed_results_count`, I get an error:

```markdown
Exception: Search plugin reported Error: Couldn't read the configuration file: Cannot parse config as TOML. Stork recieved error: unknown field `displayed_results_count`, expected one of `surrounding_word_count`, `base_directory`, `url_prefix`, `title_boost`, `stemming`,
`html_selector`, `exclude_html_selector`, `frontmatter_handling`, `files`, `break_on_file_error`, `srt_config`, `minimum_indexed_substring_length`, `minimum_index_ideographic_substring_length` for key `input` at line 772 column 1`
```

It seems that due to your solution, it’s solely possible to change the values for the so-called Stork “Input Options”. But [**Stork configuration**](https://stork-search.net/docs/config-ref) isn’t solely “Input Options”.

@lioman , what are your thoughts on making it possible to change the values of another Stork options as well for Pelican users?

Thanks.

**Comment by lioman** [comment]
I discovered the output options later and then I want to wait till this PR is merged and implement the output options in the same way.

But apparently @justinmayer hasn't found the time yet to review it. I can change it now so that in- and output options are configurable but I want to know if we can do it this way.

**Comment by Kristinita** [comment]
**Type: Another request** :mailbox_with_mail:

@lioman, You also use `stemming = "German"` in your example, but [**according to Stork `stemming` documentation**](https://stork-search.net/docs/languages) I think the `stemming` value should be generated completely automatically for the Pelican based on user Pelican settings. Example of generated `search.toml`:

```toml
[input]
base_directory = "D:/Kristinita/output"
html_selector = "section"
# [INFO] If “DEFAULT_LANG = 'ru'” in Pelican settings
# https://docs.getpelican.com/en/latest/settings.html#translations
stemming = "Russian"

# [INFO] The article on default language for the project,
# no “stemming_override”
[[input.files]]
path = "Статья-на-русском-языке.html"
url = "/Статья-на-русском-языке.html"
title = "Статья на русском языке"

# [INFO] “Lang: en” metadata in Pelican article or page:
# https://docs.getpelican.com/en/latest/content.html#file-metadata
[[input.files]]
path = "Article-in-English.html"
url = "/Article-in-English.html"
title = "Article in English"
stemming_override = "English"

```

Thanks.

**Comment by justinmayer** [comment]
Really sorry for the delay, Lioman. I'll be on vacation until the end of September but will do my utmost to review it upon my return. Many thanks for all your patience! 😅

**Comment by lioman** [comment]
> Really sorry for the delay, Lioman. I'll be on vacation until the end of September but will do my utmost to review it upon my return. Many thanks for all your patience! 😅

No problem - have a nice vacation

**Comment by lioman** [comment]
> **Type: Another request** 📬
> 
> @lioman, You also use `stemming = "German"` in your example, but [**according to Stork `stemming` documentation**](https://stork-search.net/docs/languages) I think the `stemming` value should be generated completely automatically for the Pelican based on user Pelican settings.…

I don't get what you say here. but perhaps this is for another issue to automatically detect the language based on file language. This would be another issue/pr and a different discussion.
This should not change anything here as this one is just for setting anything, that is not already covered.

I'm even willed to implement it. As this would be nicer, to have this automatically covered and no need to think about it

**Comment by paulocoutinhox** [comment]
It will be merged?

**Comment by lioman** [comment]
@justinmayer please review it - it is open so long now

**Comment by nandac** [comment]
@lioman @justinmayer I would love to see this feature merged and released. Is there any way I may help with this?

**Comment by lioman** [comment]
> @lioman @justinmayer I would love to see this feature merged and released. Is there any way I may help with this?

Me too, waiting a long time for it

**Comment by justinmayer** [comment]
Hey folks. I'm aware, and I'm doing the best I can to allocate time to review this PR this month. Thanks for your patience.

**Comment by lioman** [comment]
Thanks @justinmayer ! If I can do anything to help, just call out.

**Comment by lioman** [comment]
@justinmayer I just resolved the conflicts - hope you find some time to review this

**Comment by lioman** [comment]
@justinmayer sorry to ping you again. I just want to complete this issue. Do I need to change something or is it fine like it is?

**Comment by justinmayer** [comment]
First, thanks to @lioman for taking the time to implement and submit this pull request.

My main concern with the implementation in its current form is that I originally added two of the more important "input" options as separate settings. It seems awkward to add the rest of the Stork input options in a separate monolithic setting, with the original two dangling off by themselves. I imagine it would be better to either add all the settings as separate (which seems hard to maintain) or to move the original two separate input settings into a new monolithic setting. The latter seems like the far more sensible approach to me, of course with a deprecation period that maintains backwards compatibility for a suitable length of time.

Perhaps we could start off with new `STORK_INPUT_OPTIONS`, `STORK_OUTPUT_OPTIONS`, etc., since these settings are no longer tied to the Search plugin so much as they are to Stork itself. I think they would probably be best as key/value Python `dict` objects that are then unpacked into the TOML form that Stork expects.

Thoughts?

**Comment by lioman** [comment]
I deliberately chose to handle those input options differently, as the other two should come from the rest of config. But that was due to my simple approach of using just a plain string. With a python object I think we have more power here and I will implement it with a sensible default, so that nothing changes for the ones without changes.
I suggest, to introduce `tomli-w` as a dependency. (Just learned that `tomllib`/`tomli` can only read toml)
Is that fine @justinmayer ?

**Comment by justinmayer** [comment]
Sure, adding a dependency to more easily write TOML should be fine. As far as available options, in addition to https://github.com/hukkin/tomli-w, there is also [TOMLkit](https://github.com/sdispater/tomlkit). I've never used either, so feel free to use whichever seems like the best fit.

**Comment by lioman** [comment]
@justinmayer I have rewritten the hole stork settings generation with support for input and output options.
Additionally, I added some tests, so workflow needs an approval first. Please be so kind and review it soon.

File: .github/workflows/main.yml
Comment (justinmayer) [inline .github/workflows/main.yml:33]: Why was `--diff` removed here?

Also, please add a newline between these two steps (linters and tests).

File: .gitignore
Comment (justinmayer) [inline .gitignore]: My perspective is that most `.gitignore` directives belong on individual developers’ workstations in their global `~/.gitignore` files, leaving only the project-specific directives in project repositories like this one. Even though this project isn't using Coverage yet, I suppose the two Coverage-related files could be useful in the future. My suggestion is to change this file to the following contents:
```
.coverage
coverage
poetry.lock
```

File: README.md
Comment (justinmayer) [inline README.md]: The extra newlines removed from the README in this PR were intentionally added to improve readability when viewing/editing Markdown as raw, unrendered text (which is the way I interact with Markdown files 99% of the time).

These kinds of stylistic, preference-related changes are better discussed in a separate context, so let's revert these newline deletions.

File: README.md
Comment (justinmayer) [inline README.md]: I can't tell what change is being made here, so let's unless that can be made clear, let's revert it.

File: README.md
Comment (justinmayer) [inline README.md:50]: Let's add a newline above this sentence. Otherwise Markdown will render it as appended it to the previous paragraph.

File: README.md
Comment (justinmayer) [inline README.md]: Capital "S" `Selector`, just for consistency with other headings.

File: README.md
Comment (justinmayer) [inline README.md]: Suggested modification to language/formatting:

> To set it to a different selector (for example, `primary`), you can set it like this:

File: README.md
Comment (justinmayer) [inline README.md]: dict → `dict`

File: README.md
Comment (justinmayer) [inline README.md]: dict → `dict`

case sensitive… → case-sensitive and must be lower-case.

File: README.md
Comment (justinmayer) [inline README.md]: Unless there's a clear reason for it, let's revert this change.

**Review by justinmayer — CHANGES_REQUESTED** [review]
Nice work, Lioman! I proposed some minor suggested changes. I haven't tested the code itself but assume it works well. Let me know when everything is ready for another look. 💫

File: README.md
Comment (lioman) [reply README.md]: The reason behind this was a bug in prettier.io which is resolved in [2.8.2](https://github.com/prettier/prettier/blob/main/CHANGELOG.md#282) . Will change it back to the original one.

**Comment by lioman** [comment]
Hi @justinmayer I changed the documentation according to your remarks. If we good to go for a merge I suggest to squash the commits together as there are many - or we decline this one and I open one with this branch: https://github.com/lioman/search/tree/4-add-control-over-input-output-options

**Review by justinmayer — APPROVED** [review]
@lioman: Nice work. Thank you for addressing those minor points, some of which I tweaked further myself to save you the trouble of more back-and-forth.

Would you please add a `RELEASE.md` file in the project root, with release type of `minor`? You can see the following commit to see an example `RELEASE.md` file that was used to publish the last release: https://github.com/pelican-plugins/search/commit/cb1cb66898462d6b08af7ed8ce207ec9992ae641

**Comment by justinmayer** [comment]
@lioman: I have taken care of adding the `RELEASE.md` file and am about to merge this PR, which should trigger a new release.

As a suggestion for the future, I highly recommend doing all work on a feature/topic branch instead of the default `main` branch, even in the context of your own fork. Doing so really does help make things easier for maintainers. In this case, the release automation logic got confused by the `main` branch designation (even though it wasn't *this* repository's `main` branch) and tried to issue a release upon a push event, even though the PR hadn't been merged yet. Using a feature/topic branch would have prevented that problem.

Thank you for all your work on — and patience with — this pull request. This is a really nice enhancement to the plugin. Much appreciated! ✨

**Pull request merged**
